### ISSUE TITLE
💡 RFC: Ability to create/build custom files

### ISSUE THREAD

**Issue by Tc-001** [withastro/astro#965]
### Background & Motivation

I wish to create a image scaler at build time. It would take in an image, scale it to multiple resolutions, and return an srcset ready to use. But a roadblock is that Astro can only programmatically create HTML files.

Although it could be possible with an fs module and some trial and error, There may be many edge cases and dev/build differences that it could be easier to integrate into astro at a lower level and deal with any there.

### Possible solutions
`Astro.createFile(string | Buffer, options):{src, filepath, ...}` Takes in a buffer, and automatically creates a cusom, collision-free(if no filename specified) file from it, and returns the public file path (as accessed from the browser)
```js
Astro.createFile(some_modofied_file, {extension: "png"}).src //   "/_astro/Component/H86jd.png"
Astro.createFile(some_modofied_file, {name: "image5", extension: "png"}).src //   "/_astro/Component/image5.png"
Astro.createFile(some_modofied_file, {
    path: "/api",
    name: "allposts", 
    extension: "json"
}).src //   "/api/allposts.json"
```
This could also be used as a "dummy" file for other libraries that require a file to be written to
```js
//          ./src/pages/News.astro
const path = Astro.createFile().filepath //  "/home/user/.../H86jd"
some_lib().write(path)
```
### Alternatives considered
* Something similar to SvelteKit's endpoints, where you have a custom .ts (ex. `$scale.ts`), that when called expects a buffer as the return. In that case, there are issues with how to return the filepath to the .astro file, as there would be a disconnect between what is in the endpoint, and what is being imported.
```js
export default function rescale() {
   //magic
   return buffer
}
```
* Previous, but the return value is an object, then it would be similar to the collection api or sveltekit's load
* Astro only suggests a filename for it, user has to actually create it. This is more low-lever, but prone to user error (also may be harder for the dev server to deal with):
 ```js
const path = Astro.allocateFile(...); //   "./astro/H86jd.png" or maybe "/absolute/path/(dist|public|temp|whatever)/astro/H86jd.png"
fs.writeFile(..., path, ...);
```
### Risks, downsides, and/or tradeoffs
* Theoretically it would be possible to create malicious files, if there are not proper checks in place and the database the build is pulling from gets compromised (`evil.php` as an image source, for example)
* Easy to mess up JS buffers and get stuck (additional issue/help message volume)
* If a webmaster manually copies files through FTP after the build (and does not clear out the directory), they could end up with multiple copies of the same, or slightly changed files, especially if they use any kind of dynamic or generated image.
### Open Questions
* What should be the default path, will the user be able to change it? 
* How will the dev server handle this? (will it re-run for each request, or cache the result somewhere)
* The user will probably want to specify a lot of parameters. Would it be better to bundle everything in an `options` object? `Astro.createFile({buffer: foo, filename: bar...})`
* How should it handle attempting to override an astro-built file? (ex. `index.html` in webroot)

### Help make it happen!

- [ ] I am willing to submit a PR to implement this change.
- [X] I am willing to submit a PR to implement this change, but would need some guidance.
- [ ] I am not willing to submit a PR to implement this change.

**Comment by Tc-001** [comment]
Poking around the discord more, I see that it may be needed for the image efforts already taking place at `#🏗🎨-images`. But I am not sure if this would help them, or if someone already figured everything out this far.

I believe setting this as a simple API will help with maintaining the image side of things easily readable, while also allowing everyone to use it for different things in the future. (custom `.htacess` comes to mind)

**Comment by jasikpark** [comment]
I definitely think figuring out how to program and generate files and save them to disk for inclusion in the website is the first step to getting images support.￼

**Comment by mash-graz** [comment]
handling the file access is indeed one of most challenging aspects of this feature request.
although i generally sympathize with the suggested improvements, i also see the danger of hardening those issues, which already render the utilization of astro in deno and pure browser based IDEs impossible (see: #566).

**Comment by Tc-001** [comment]
Depending on the environment there could be different implementations. 
For a static site they get generated in `/dist/` with everything else, while for deno or browser IDEs, the resulting source path could point to an endpoint of some sort. 
So maybe `Astro.renderFile()` would be a better name

**Comment by Tc-001** [comment]
Maybe a good name (should help with keeping it browser-like) would be (createObjectURL)[https://developer.mozilla.org/en-US/docs/Web/API/URL/createObjectURL]

**Comment by Tc-001** [comment]
Sorry, the worst missclick imaginable

**Comment by mash-graz** [comment]
> Maybe a good name (should help with keeping it browser-like) would be Astro.createObjectURL()

yes -- it has to be something more abstract, although most runtimes support one or the other way to access at least some kind of virtual file system...

*"We encourage Astro users to avoid Node.js builtins (`fs`, `path`, etc) whenever possible. Astro aims to be compatible with multiple JavaScript runtimes in the future. This includes [Deno](https://deno.land/) and [Cloudflare Workers](https://workers.cloudflare.com/) which do not support Node builtin modules such as `fs`." ([docs/src/pages/guides/imports.md](https://github.com/snowpackjs/astro/blob/6300c88cacc09f8b947c531662c25ac77166dfe7/docs/src/pages/guides/imports.md))*

but the main problem doesn't concern the naming conventions or utilized compatibility layers but the fact, that we are not able to sufficiently detect and check this kind of incompatibilities affecting other runtimes as long as core parts of astro only work in node envrionments.

**Comment by Tc-001** [comment]
So for now I will focus my implementation on nodejs with fs access for building and a static site for prod.
In the future it could expand, so I will try and make it as modular as possible :+1:

**Comment by Tc-001** [comment]
I have a proof of concept that just creates the file in a `./public/astro_custom_files` folder (temp jank to test if it is even possible to do)

But figuring out how to make it integrate well is proving to be a challenge, because it needs to interact with both runtime JS and snowpack.
I was hoping to make it a separate renderer, but from the looks of it, it probably needs to integrate in astro itself for this to work well.

**Comment by Tc-001** [comment]
Going by what FredKSchott said on discord (about https://github.com/snowpackjs/astro/pull/1010)
> Oh actually great call, that argument isn’t documented yet. In the new API, it’s rss({…}, {url: “/your/custom/feed.xml”)
[source](https://discord.com/channels/830184174198718474/830184175176122389/873644902590844966)

So whenever that gets merged, astro will have functionality to create files in custom paths (the main thing I have been struggling with while making a proof of concept). So that could just be extended to more arbitrary files like plaintext or images.

**Comment by drwpow** [comment]
This is really interesting! This is actually something Fred and I have been thinking about all during the development of Snowpack: _How can you create a general interface for building files?_ Unfortunately in practice, I think this question is really specific to _what_ you’re generating, as you’ve pointed out. You have different considerations and needs for images than you do RSS feeds.

Though this is a question I’ve thought about a lot myself, every time I think about a generalized builder interface I eventually accidentally recreate [Gulp.js](https://gulpjs.com/), which is something I don’t want to do 😅.

But that said, I’m always in favor of more hooks in Astro. I know this is different than what’s proposed, but what if we had some JS API that you could tap into as part of the build? (this is all pseudo-code; don’t get hung up on the syntax; just trying to illustrate a rough idea) 

```
import astro from 'astro';

const builder = new astro.Builder('./');
await builder.build(); // start build

builder.on('load', (filename, contents, finalDest) => {
  // do something with filename, contents, and finalDest path
});
```

Again, this is 100% made up and can be improved; just trying to illustrate some way you could see all files that Astro is processing along with their contents, and use it to write more files to disk or do whatever you’d like.

Of course, this leaves out dev, which currently that would be very hard to do, but with some changes to the new compiler that will be easier, so let’s just say wait on that part for now.

What are your thoughts on this approach? Is this a bit different from your original usecase?

**Comment by Tc-001** [comment]
That could be good, but I see it more like this (going by your example)
```js
import astro from 'astro';

const file = new astro.File('./foo.txt');
await file.insert("Content");// add text to file
```
That could be added to any page-level astro file. (Generated file being inserted either in the folder next to `index.html` or in `_astro/`)

Because I think while build hooks are good, they in my opinion are not a end-all solution, because it would be hard for a newcomer to pick them up.

So mybe the hooks could be used under the hood, but with some kind of abstraction API on top.

That will probably be what will happen either way, but just want to clear that up.

**Comment by Tc-001** [comment]
Looking over it again, I think this is the way to go.
Because if the low level stuff can be implemented, we should be able to just port https://github.com/JonasKruckenberg/imagetools or similar library. 

Maybe it is more snowpack, but if someone could write a package (guiding from these: https://github.com/JonasKruckenberg/imagetools/tree/main/packages) then the imagetools library should just work (I hope)

**Comment by FredKSchott** [comment]
RFC Call:
- Docs on SvelteKit: https://kit.svelte.dev/docs#routing-endpoints
- Image RFC: https://github.com/snowpackjs/astro/issues/492
- No consensus reached, but definitely interested in solving the image use-case! 
- Recommendation: focusing on Image component use-case
- Alternative mentioned: SvelteKit endpoint-style support - interest in a dedicated RFC as well!

**Comment by tony-sull** [comment]
[Link to the SvelteKit's endpoint implementation](https://github.com/sveltejs/kit/blob/02f7aba5ea88f8f423d626b4c430be4760bcc561/packages/kit/src/runtime/server/endpoint.js#L41)

With the focus here being on the image use case, that may be less important with `vite-imagetools` making it's way into Astro land when the new compiler is released.

Maybe it would be worth tabling this RFC for now, and revisiting it down the road?  I could see this being extremely useful if serverless rendering (edge builders, etc.) becomes a priority for Astro, but less useful other than image transformations as long as Astro is focused only on SSG

**Comment by Tc-001** [comment]
Sounds good to me! 
Especially with #1305 covering most of the other usecases

**Comment by FredKSchott** [comment]
Hey everyone! Our current RFC process is beginning to break down at this size, with over 50 open RFCs currently in the "discussing" stage. A growing community is a great problem to have, but our ability to give you RFC feedback has suffered as a result. In an effort to improve our RFC process, we are making some changes to better organize things. 

**From now on, all RFCs will live in a standalone repo:** https://github.com/withastro/rfcs

This allows us to do three things: 1) Use threaded discussions for high-level ideas and improvements, without necessarily requiring an implementation for every idea. 2) Improve the quality of our RFC template and the speed/quality of all feedback. 3) Support inline comments and explicit approvals on RFCs, via a new Pull Request review process.

We hope that this new process leads to better RFC weekly calls and faster feedback on your RFCs from maintainers. More detail can be found in the new RFC repo [README](https://github.com/withastro/rfcs).

--- 

**We can't automatically convert this issue to an RFC in the new repo** because new RFC template is more detailed that this one. But, you can still continue this discussion in the new repo by [creating a new Discussion](https://github.com/withastro/rfcs/discussions) in the RFC repo and copy-and-pasting this post (and any relevant follow-up comments) into it. Discussions are available for high-level ideas and suggestions without the requirement of a full implementation proposal.

Then, when you are ready to propose (or re-propose) an implementation for feedback and approval, you can create a new RFC using the new RFC template. More detail about how to do this can be found in the new RFC repo [README](https://github.com/withastro/rfcs).

Thanks for your patience as we attempt to improve things for both authors and reviewers. If you have any questions, don't hesitate to reach out on Discord. https://astro.build/chat
